### PR TITLE
Feature/client id and firebase app

### PR DESCRIPTION
**Pull request by rptomey**
Support for tracking using either gtag or Firebase, including the creation of new subclasses, updates to the markdown, and a utility function for generating a random new client id (gtag only).

### CODE REVIEW

File: ga4mp/ga4mp.py [hunk 1]
```diff
@@ -20,18 +21,13 @@
 
 class Ga4mp(object):
```
Comment (zvizdo) [inline ga4mp/ga4mp.py]: Would name BaseGA4MP to denote that it is a base class and not something you'd make an instance of to use the package.

File: ga4mp/ga4mp.py [hunk 2]
```diff
@@ -343,3 +343,61 @@ def _check_date_not_in_future(self, date):
             assert (
                 date <= datetime.datetime.now()
             ), "Provided date cannot be in the future"
+
+class gtagMP(Ga4mp):
```
Comment (zvizdo) [inline ga4mp/ga4mp.py]: Capitalize

File: ga4mp/ga4mp.py [hunk 3]
```diff
@@ -343,3 +343,61 @@ def _check_date_not_in_future(self, date):
...
+    def build_request(self, batch):
+        return {"client_id": self.client_id, "events": batch}
+
+    def random_client_id(self):
+        """
+        Utility function for generating a new client ID matching the typical format of 10 random digits and the UNIX timestamp in seconds, joined by a period.
+        """
+        return "%0.10d" % random.randint(0,9999999999) + "." + str(int(time.time()))
+
+class firebaseMP(Ga4mp):
```
Comment (zvizdo) [inline ga4mp/ga4mp.py]: Capitalize

File: ga4mp/ga4mp.py (hunk 1 above)
Comment (zvizdo) [reply ga4mp/ga4mp.py]: Include 
_build_url, _build_request

Something line:
```
def _build_url(self, domain):
    raise NotImplementedError()
```

File: ga4mp/ga4mp.py [hunk 4]
```diff
@@ -343,3 +343,61 @@ def _check_date_not_in_future(self, date):
...
+    client_id : string
+        A unique identifier for a client, representing a specific browser/device.
+    """
+
+    def __init__(self, api_secret, measurement_id, client_id):
+        super().__init__(api_secret)
+        self.measurement_id = measurement_id
+        self.client_id = client_id
+
+    def build_url(self, domain):
```
Comment (zvizdo) [inline ga4mp/ga4mp.py]: Overwrite the base class...

File: ga4mp/ga4mp.py [hunk 5]
```diff
@@ -343,3 +343,61 @@ def _check_date_not_in_future(self, date):
...
+        return f"{domain}?measurement_id={self.measurement_id}&api_secret={self.api_secret}"
+
+    def build_request(self, batch):
```
Comment (zvizdo) [inline ga4mp/ga4mp.py]: Overwrite the base class...

File: ga4mp/ga4mp.py [hunk 6]
```diff
@@ -343,3 +343,61 @@ def _check_date_not_in_future(self, date):
...
+            * C++ - GetAnalyticsInstanceId() - https://firebase.google.com/docs/reference/cpp/namespace/firebase/analytics#getanalyticsinstanceid
+            * Unity - GetAnalyticsInstanceIdAsync() - https://firebase.google.com/docs/reference/unity/class/firebase/analytics/firebase-analytics#getanalyticsinstanceidasync
+    """
+
+    def __init__(self, api_secret, firebase_app_id, app_instance_id):
+        super().__init__(api_secret)
+        self.firebase_app_id = firebase_app_id
+        self.app_instance_id = app_instance_id
+
+    def build_url(self, domain):
```
Comment (zvizdo) [inline ga4mp/ga4mp.py]: Overwrite the base class...

File: ga4mp/ga4mp.py [hunk 7]
```diff
@@ -343,3 +343,61 @@ def _check_date_not_in_future(self, date):
...
+        return f"{domain}?firebase_app_id={self.firebase_app_id}&api_secret={self.api_secret}"
+
+    def build_request(self, batch):
```
Comment (zvizdo) [inline ga4mp/ga4mp.py]: Overwrite the base class...

File: ga4mp/ga4mp.py (hunk 3 above)
Comment (rptomey) [reply ga4mp/ga4mp.py]: Done, and I updated it in the markdown for the readme, too.

File: ga4mp/ga4mp.py (hunk 2 above)
Comment (rptomey) [reply ga4mp/ga4mp.py]: Done, and I updated it in the markdown for the readme, too.

File: ga4mp/ga4mp.py (hunk 1 above)
Comment (rptomey) [reply ga4mp/ga4mp.py]: Updated the function names and the references to include the preceding underscore, then added these two functions to the base class:
```    
def _build_url(self, domain):
    raise NotImplementedError("Subclass should be using this function, but it was called through the base class instead.")

def _build_request(self, batch):
    raise NotImplementedError("Subclass should be using this function, but it was called through the base class instead.")```

File: ga4mp/ga4mp.py (hunk 4 above)
Comment (rptomey) [reply ga4mp/ga4mp.py]: Handled as suggested!

File: ga4mp/ga4mp.py (hunk 6 above)
Comment (rptomey) [reply ga4mp/ga4mp.py]: Handled as suggested!

File: ga4mp/ga4mp.py (hunk 5 above)
Comment (rptomey) [reply ga4mp/ga4mp.py]: Handled as suggested!

File: ga4mp/ga4mp.py (hunk 7 above)
Comment (rptomey) [reply ga4mp/ga4mp.py]: Handled as suggested!

**Pull request merged**